cr-examples/onnx/src/main/java/oracle/code/onnx/OnnxRuntime.java

```diff
@@ -0,0 +1,535 @@
+
+    public Environment createEnv(LoggingLevel logLevel, String logId) {
+        try {
+            return new Environment(retAddr(createEnv.invokeExact(logLevel.ordinal(), arena.allocateFrom(logId), ret)),
+                                   retAddr(getAllocatorWithDefaultOptions.invokeExact(ret)));
+        } catch (Throwable t) {
+            throw wrap(t);
+        }
+    }
+
+    public enum LoggingLevel {
+        VERBOSE, INFO, WARNING, ERROR, FATAL
+    }
+
+    public enum ONNXType {
+        UNKNOWN, TENSOR, SEQUENCE, MAP, OPAQUE, SPARSETENSOR, OPTIONAL
+    }
+
+    public final class Environment {
+
+        private final MemorySegment envAddress,
+                                    defaultAllocatorAddress;
+
+        private Environment(MemorySegment envAddress, MemorySegment defaultAllocatorAddress) {
+            this.envAddress = envAddress;
+            this.defaultAllocatorAddress = defaultAllocatorAddress;
+        }
+
+        public List<OrtTensor> runOp(OnnxOp.OnnxSchema schema, List<OrtTensor> inputValues) {
+            var inputs = schema.inputs();
+            // @@@ hardcoded float tensor element type
+            try (var session = createSession(OnnxProtoBuilder.opModel(schema, Tensor.ElementType.FLOAT))) {
+                return session.run(
+                        IntStream.range(0, inputs.size()).boxed().collect(Collectors.toMap(i -> inputs.get(i).name(), i -> inputValues.get(i))),
+                        schema.outputs().stream().map(OnnxOp.OnnxParameter::name).toList());
+            }
+        }
+
+        public Session createSession(String modelPath) {
+            return createSession(modelPath, createSessionOptions());
+        }
+
+        public Session createSession(String modelPath, SessionOptions options) {
+            try {
+                return new Session(retAddr(createSession.invokeExact(envAddress, arena.allocateFrom(modelPath), options.sessionOptionsAddress, ret)));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+        }
+
+        public Session createSession(ByteBuffer model) {
+            return createSession(model, createSessionOptions());
+        }
+
+        private Session createSession(ByteBuffer model, SessionOptions options) {
+            try {
+                return new Session(retAddr(createSessionFromArray.invokeExact(envAddress, MemorySegment.ofBuffer(model.rewind()), (long)model.limit(), options.sessionOptionsAddress, ret)));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+        }
+
+        public final class Session implements AutoCloseable {
+
+            private final MemorySegment sessionAddress;
+
+            private Session(MemorySegment sessionAddress) {
+                this.sessionAddress = sessionAddress;
+            }
+
+            public int getNumberOfInputs() {
+                try {
+                    return retInt(sessionGetInputCount.invokeExact(sessionAddress, ret));
```

Comment:
I note the use of this shared `ret` output segment. We plan to add more capabilities to FFM to address the case of recyclable allocation -- this will help code like this to avoid the shared segment, at the same time while avoiding the cost of a malloc per call.

Reply:
That would be very helpful.
Truly I haven't found any FFM hints which type of arena is best for specific case.
My guess is that garbage-collected allocations may be better than creating and disposing individual arena for each allocation (I've seen a lot of objects constructed).
Simple manual disposal of a memory segment might be a step forward?

Reply:
Finally I've decided to re-use the `ret` as the thread-safety is not yet a problem.